### ISSUE TITLE
Support relocatable packages

### ISSUE THREAD

**Issue by bos** [haskell/cabal#462]
(Imported from [Trac #469](http://hackage.haskell.org/trac/hackage/ticket/469), reported by @dcoutts on 2009-01-20)

This is both useful and doable on both Windows and Unix systems.

We already have limited support for relocatable (also sometimes called prefix-independent) packages on Windows. Specifically, a relocatable package is often what people want when they want to prepare a package for redistribution, especially on Windows for use with an installer. What is required is that an exe and all supporting files (data files and potentially also other shared libs) exist together in one directory hierarchy and that that directory can be placed anywhere in the filesystem and the exe will still be able to find all its associated support files.

There are some restrictions on preparing a relocatable package. In particular the user must configure it such that all install directories for the various kinds of files (libdir, datadir etc) are relative to the <tt>$prefix</tt>.

Configuring a relocatable package should be something that is done explicitly. At the moment it is done simply by configuring it in the right way, but Cabal is never aware that the user is trying to construct a relocatable package. If it is explicit then we can take different actions if necessary (as it is on unix) and do additional checks.

So a configure option <tt>--relocatable</tt> or something should be used. Installing a relocatable package is somewhat of an exception because it involves using a prefix which would otherwise be empty. Creating an image for a relocatable package would use a empty <tt>$prefix</tt>.

The tricky issues that have to be addressed are:

<ul><li>data files for the program itself
</li><li>data files in library dependencies
</li><li>shared libraries
</li></ul>The mechanism for finding data files at runtime is the <tt>Paths_pkgname</tt> module that Cabal generates. On Windows this can make use of the Win32 api that lets a .exe program discover where it was run from. On unix there is no reliable equivalent but the next best alternative is to use wrapper scripts to set environment variables before running the real program.

This mechanism can work pretty well for data files that belong to the executable. It is rather harder for data files in dependent packages (see #459). One option would be to copy the data files belonging to dependent packages and install them along with the executable (checking for clashes). This would involve being able to locate the data files of installed libraries which would require extending the data stored for installed packages. Libraries that require data files are relatively rare (though this should be checked) so this feature could be punted for an initial version.

In future there will also be the issue of shared libraries to consider. That is, when Haskell implementations can produce Haskell libs as shared libraries. This complicates the construction of relocatable packages. Again the approach taken on Windows and Unix will have to be different. On Windows it will be possible to copy the .dll files into the same directory as the .exe file (or a subdir if using .manifest files). On Unix a wrapper script using <tt>LD_LIBRARY_PATH</tt> will probably be necessary.

**Comment by mietek** [comment]
@christiaanb: Does #2255 being merged mean this issue can be closed?

**Comment by christiaanb** [comment]
No, this issue cannot be closed. At the end of #2255 it says:

As said, this partially implements #462. To fully implement that issue, we would additionally need a deploy command. Which you would call like: cabal deploy --deploy-dir=<deploy_dir> <package> which would collect the binaries, libraries of the dependencies (in case of a dynamically linked executable), and data-files, and turn them into a relocatable bundle at the specified directory.

**Comment by mietek** [comment]
@christiaanb: From the point of view of the user, what would be the difference between `cabal deploy --deploy-dir=...` and `cabal install --prefix=...`?

**Comment by christiaanb** [comment]
@mietek: `cabal deploy` is meant for a developer who wants to package his application for distribution to users. It would be equivalent to an `.app` bundle on OS X.  `cabal install` is only something that's meant to be used by developers, not by end-users of an application that happens to be written in Haskell.

**Comment by BardurArantsson** [comment]
@christiaanb Are you still working on this and (if so) how close is this to landing?

**Comment by christiaanb** [comment]
@BardurArantsson No, #2255 is the last I worked on it. #2255 provides a good start, I'd say it gets you 33% to 50% of achieving the goals of this (#462) issue. What is still needed:

* A better story for `data-files`, see: https://github.com/haskell/cabal/pull/2255#issuecomment-66086160
* A `cabal deploy` command which creates something like an OS X `.app` bundle.
* Update GHC so that the `base` packages are fully relocatable.

I don't see myself working on the above issues any time soon. What #2255 does get you is (mostly) relocatable sandboxes: http://christiaanb.github.io/posts/relocatable-sandboxes/

**Comment by mantkiew** [comment]
One way people (including myself) deal around the `data-files` issue is by using the [file-embed](https://hackage.haskell.org/package/file-embed) package. I could imagine `cabal-install` use `file-embed` for every file in `data-files` and make them available via the `Package_xyz.hs` module. When you use `file-embed`, you get a `[(filePath, contents)]` list. This way, all data files would be directly embedded within the executable.

**Comment by christiaanb** [comment]
@mantkiew: what's the cross-platform support for `file-embed`, and how well does it scale? as in, is it "OK" to embed a 100 MB, or even 1GB, data-file in the executable?

**Comment by mantkiew** [comment]
I confirm that it works on Windows and Linux. I haven't tested on MacOS yet. It certainly is not a universal mechanism and there will be size limits. Besides, keeping 1GB on the heap does not make any sense. I only used it for small files.

**Comment by essandess** [comment]
@bos @simonmar @christiaanb 
At MacPorts we’ve [automated](https://github.com/macports/macports-ports/blob/master/_resources/port1.0/group/haskell_stack-1.0.tcl) `stack` builds and this has been successful for packages that don’t use `datadir`.

However, `cabal`’s lack of a GNU-standard capability for `DESTDIR` and `PREFIX` to produce relocatable binaries and installs is causing major problems for package managers that need to build with Haskell development tools. E.g. see:
* https://github.com/macports/macports-ports/pull/4706
* https://github.com/ndmitchell/hlint/issues/699
* https://github.com/macports/macports-ports/pull/5050#issuecomment-527435346
* https://github.com/macports/macports-ports/pull/5180

There’s obviously a lot of interest in the capability and the lack of it is holding back deployment of Haskell-based tools.

Are there any plans to tackle this soon?

Related:
* https://github.com/haskell/cabal/issues/462
* https://github.com/haskell/cabal/issues/3586
* https://github.com/commercialhaskell/stack/issues/848
* https://github.com/commercialhaskell/stack/issues/4857

**Comment by 23Skidoo** [comment]
No one is actively working on this AFAIK, but I think it'd make a good GSoC project.

**Comment by essandess** [comment]
The [approach](http://neilmitchell.blogspot.com/2008/02/adding-data-files-using-cabal.html) described by [hlint](https://github.com/ndmitchell/hlint)'s author @ndmitchell solves the issue of hardcoded `cabal` `data-files` in the binary.

All that's required is to specify the path within a file called `Paths_packagename.hs`, which is normally automatically generated by `cabal` with its own paths. Here is an automated `stack` build that solves the issue along with related files:
* https://github.com/macports/macports-ports/blob/e18fe88466477b20da595346ebce8eab839b9fbc/www/adblock2privoxy/Portfile
* https://github.com/macports/macports-ports/blob/e18fe88466477b20da595346ebce8eab839b9fbc/www/adblock2privoxy/files/Paths_adblock2privoxy.hs

Also, setting the environment variable `packagename_datadir` at runtime overrides the binary's hardcoded `packagename_datadir` path.

Also see https://github.com/commercialhaskell/stack/issues/5026.
